src/number_theory/fermat_psp.lean

```diff
@@ -0,0 +1,578 @@
+    rw nat.sub_sub_self q₂ at this,
+    exact nat.prime.not_dvd_one hk this },
+
+  -- If n = 1, then it follows trivially that n is coprime with b.
+  { have : n = 1 := by linarith,
+    rw this,
+    norm_num }
+end
+
+lemma probable_prime_iff_modeq (n : ℕ) {b : ℕ} (h : 1 ≤ b) :
+  probable_prime n b ↔ b^(n - 1) ≡ 1 [MOD n] :=
+begin
+  have : 1 ≤ b ^ (n - 1) := one_le_pow_of_one_le h (n - 1), -- For exact_mod_cast
+  split,
+  { intro h₁,
+    apply nat.modeq_of_dvd,
+    have h₂ : ↑n ∣ ↑(b ^ (n - 1)) - (1 : ℤ) := by exact_mod_cast h₁,
+    have h₃ : - (↑(b ^ (n - 1)) - (1 : ℤ)) = ((↑1 : ℤ) - ↑(b ^ (n - 1))) := by simp,
+    exact h₃ ▸ (dvd_neg _ _).mpr h₂ },
+  { intro h₁,
+    have h₂ : ↑n ∣ ↑(b ^ (n - 1)) - (1 : ℤ) := nat.modeq.dvd h₁.symm,
+    exact_mod_cast h₂ }
+end
+
+/--
+If `n` is a Fermat pseudoprime to base `b`, then `n` is coprime with `b`, assuming that `b` is
+positive.
+
+This lemma is a small wrapper based on `coprime_of_probable_prime`
+-/
+lemma coprime_of_fermat_psp (n b : ℕ) (h : fermat_psp n b) (h₁ : 1 ≤ b) : nat.coprime n b :=
+begin
+  cases h with hp hn,
+  cases hn with hn₁ hn₂,
+  exact coprime_of_probable_prime n b hp (by linarith) h₁,
+end
+
+/--
+All composite numbers are Fermat pseudoprimes to base 1.
+-/
+lemma pseudoprime_of_base_one (n : ℕ) (h₁ : 1 < n) (h₂ : ¬nat.prime n) : fermat_psp n 1 :=
+begin
+  split,
+  { have h : 0 = 1^(n - 1) - 1 := by norm_num,
+    show n ∣ 1^(n - 1) - 1, from h ▸ (dvd_zero n) },
+  { exact ⟨h₂, h₁⟩ }
+end
+
+-- For lemmas that are needed to prove statements in this file, but aren't directly related to
+-- Fermat pseudoprimes
+section helper_lemmas
+
+private lemma odd_of_prime_gt_two (p : ℕ) (h : nat.prime p) (hp : 2 < p) : odd p :=
+begin
+  rw [nat.odd_iff_not_even, even_iff_two_dvd],
+  rw nat.prime_dvd_prime_iff_eq nat.prime_two h,
+  exact ne_of_lt hp
+end
+
+private lemma odd_pow_lem (a : ℤ) (n k : ℕ) (h₁ : k ∣ n) (h₂ : odd (n / k)) : a^k + 1 ∣ a^n + 1 :=
+begin
+  -- Let m be the natural number such that k * m = n. Then (-1)^m = -1 since m is odd by hypothesis.
+  generalize h₃ : n / k = m,
+  have hm : k * m = n := by { rw [←h₃, mul_comm], exact nat.div_mul_cancel h₁ },
+  have hm₁ : (-1 : ℤ)^m = -1 := odd.neg_one_pow (h₃ ▸ h₂),
+
+  -- a^k ≡ -1 (mod a^k + 1)
+  have hk : a^k + 1 ∣ a^k + 1 - 1 -(-1) := by norm_num,
+  have hk₁ : a^k + 1 - 1 ≡ -1 [ZMOD a^k + 1] := (int.modeq_of_dvd hk).symm,
+  have hk₂ : a^k ≡ -1 [ZMOD a^k + 1] := by rwa int.add_sub_cancel at hk₁,
+
+  -- a^n = (a^k)^m ≡ (-1)^m = -1 (mod a^k + 1)
+  have ha : a^n = (a^k)^m := by rw [←pow_mul, hm],
+  have ha₂ : (a^k)^m ≡ (-1)^m [ZMOD a^k + 1] := int.modeq.pow m hk₂,
+  have ha₃ : a^n ≡ (-1)^m [ZMOD a^k + 1] := (eq.symm ha) ▸ ha₂,
+  have ha₄ : a^n ≡ (-1) [ZMOD a^k + 1] := hm₁ ▸ ha₃,
+
+  -- Therefore, a^k + 1 divides a^n + 1
+  show a^k + 1 ∣ a^n + 1, from (int.modeq.symm ha₄).dvd
+end
+
+private lemma sub_one_dvd_pow_sub_one (a n : ℕ) (h : a ≥ 1) : a - 1 ∣ a^n - 1 :=
```

Comment:
```suggestion
private lemma sub_one_dvd_pow_sub_one (a n : ℕ) (h : 1 ≤ a) : a - 1 ∣ a^n - 1 :=
```
Always `≤`, never `≥` (or else we'd double the number of lemmas we have about inequalities). Again I feel like this should be somewhere higher up the import tree and not private. It's also true when a = 0 I guess, so probably you should prove the more general thing (but do a case split on a first).

Reply:
Using `≥` must have been a mistake, since I thought I'd flipped all of them.
This lemma always felt like it was too general for `fermat_psp.lean`. Should I try to get this into a different part of mathlib? I have a proof of the more general `lemma ab_lem (a b n : ℕ) : (a - b) ∣ (a^n - b^n)` [here](https://github.com/osbourn/fermat-pseudoprimes/blob/99bba0e1a33779abc8d4cb0c02c26f25f89fc0f5/src/fermat_psp.lean#L178), but removed it because it was more powerful than I actually needed and relied on natural number subtraction.

Reply:
`sub_one_dvd_pow_sub_one` has been replaced by a lemma added in #17996